CODING.md

```diff
@@ -0,0 +1,97 @@
+# Coding guide
+
+This project is written in Go and adheres to the accepted coding styles and practices defined in [Effective Go](https://golang.org/doc/effective_go.html) and [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments). 
+
+## Static code analysis
+
+Linting is done by using https://github.com/golangci/golangci-lint with configuration defined in the root of the project.
+
+## Testing
+
+It is advisable to use a separate test package to write tests (e.g. with the `package somepkg_test` namespace). This allows explicit testing on exported behaviour of the package, not the internal implementation (also known as black-box-testing). In need to export only to testing package, `export_test.go` file should be used. Testing internal implementations is discouraged, but not forbidden. It is preferred to treat exported package behaviours as units that need to be tested.
+
+Executing `make test` should pass without any warnings or errors.
+
+## Packages
+
+Go packages with the spacemesh project should have a single and well defined responsibility, clear exposed API and tests that cover expected behaviour. To ensure better modularity of the codebase, every package should be treated as a library that provides a specific functionality and that can be used as a module in other applications. Every package should have a well written godoc page which should be used as the entry point for understanding the package's responsibility. The same as using any other third-party package. If the package godoc is not clear and requires looking at the code to understand behaviour that it provides, documentation should be improved.
+
+## Concurrency
+
+For every goroutine that is created, one must define how and when it terminates.
```

Comment:
So `errgroup` helps with error propagation from goroutines, but it does not define how a goroutine lifecycle should be managed. It has an optional context that can be passed, but finally it is up to the developer to decide how the goroutine lifecycle should be managed. This specific comment is made in order to clarify the conditions that must be met in order for a goroutine to terminate (it might also not even emit an error as of such, as some goroutines are just anonymous functions that return nothing). I agree that we can add that `errgroups` should be used, however, these are two separate issues.

Reply:
This is true but if you take a look at how we usually use err group you'll see that we are actually not using it for the error propagation or synchronized cancelling in most cases. But rather because

```go
eg.Go(func() error {
   // do stuff
   return nil
})

// Other calls to eg.Go then
eg.Wait()
```
Is syntactically nicer that
```go
var wg sync.Waitgroup
wg.Add(1)
go func () {
   defer wg.Done()
   // do stuff 
}()

// Other go routines with wg.add and wg.done 
wg.Wait()
```

Reply:
Most components also have an `errgroup.Group` that is used to start to routines that do not need to be awaited generally, but will be awaited when shutting down (gracefully exit).

It's a pattern that we started following some time ago and it proved to be useful.

Reply:
The `errgroup` is not only nicer to the eye, but also easier to use than waitgroup (can't misplace/forget `wg.Add(1)`).

Reply:
I totally agree that using `errgroup` is favorable to using `sync.WaitGroup`, however, this is again not what the comment is referring to. The topic is rather a bit more abstract and is mentioned in the [CodeReviewComments](https://go.dev/wiki/CodeReviewComments#goroutine-lifetimes) resource:
```
Goroutine Lifetimes
When you spawn goroutines, make it clear when - or whether - they exit.

Goroutines can leak by blocking on channel sends or receives: the garbage collector will not terminate a goroutine even if the channels it is blocked on are unreachable.

Even when goroutines do not leak, leaving them in-flight when they are no longer needed can cause other subtle and hard-to-diagnose problems. Sends on closed channels panic. Modifying still-in-use inputs “after the result isn’t needed” can still lead to data races. And leaving goroutines in-flight for arbitrarily long can lead to unpredictable memory usage.

Try to keep concurrent code simple enough that goroutine lifetimes are obvious. If that just isn’t feasible, document when and why the goroutines exit.
```

Reply:
I would like to address all of the specifics on favorable patterns and do's and don't's in another document which would be effectively the style-guide, and keep this one more abstract